crates/circuit/src/instruction.rs

```diff
@@ -0,0 +1,436 @@
+///
+/// Provides ergonomic views of an underlying instruction.
+pub trait IntoInstructionView<'a> {
+    /// The type of inner circuits contained within this instruction's views.
+    type Block;
+
+    /// Returns a view of the operation.
+    fn view_operation(self) -> OperationRef<'a>;
+
+    /// Returns a view of this instruction as a standard gate, if applicable.
+    fn try_view_standard_gate(self) -> Option<StandardGateView<'a>>;
+
+    /// Returns a view of this instruction as a standard instruction, if applicable.
+    fn try_view_standard_instruction(self) -> Option<StandardInstructionView<'a>>;
+
+    /// Returns a view of this instruction as a control flow instruction, if applicable.
+    fn try_view_control_flow(self) -> Option<ControlFlowView<'a, Self::Block>>;
+
+    /// Returns the old-style [Param] sequence, unless this is a control
+    /// flow instruction.
+    fn try_legacy_params(self) -> Option<&'a [Param]>;
+
+    /// Returns an immutable ergonomic view of this instruction.
+    #[inline]
+    fn view(self) -> InstructionView<'a, Self::Block>
+    where
+        Self: Copy + Sized,
+    {
+        match self.view_operation() {
+            OperationRef::ControlFlow(_) => {
+                InstructionView::ControlFlow(self.try_view_control_flow().unwrap())
+            }
+            OperationRef::StandardGate(_) => {
+                InstructionView::StandardGate(self.try_view_standard_gate().unwrap())
+            }
+            OperationRef::StandardInstruction(_) => {
+                InstructionView::StandardInstruction(self.try_view_standard_instruction().unwrap())
+            }
+            OperationRef::Gate(g) => InstructionView::Gate(g),
+            OperationRef::Instruction(i) => InstructionView::Instruction(i),
+            OperationRef::Operation(o) => InstructionView::Operation(o),
+            OperationRef::Unitary(u) => InstructionView::Unitary(UnitaryGateView(u)),
+        }
+    }
+}
+
+/// Represents an instruction that is directly convertible to our Python API
+/// instruction type.
+pub trait Instruction {
+    /// Gets a reference to this instruction's operation.
+    fn op(&self) -> OperationRef<'_>;
+
+    /// Get a reference to this instruction's parameter list, if applicable.
+    ///
+    /// For standard gates without parameters this may be [None] or a
+    /// `Some(Parameters::Param(smallvec![]))`.
+    fn parameters(&self) -> Option<&Parameters<PyObject>>;
+
+    /// Get the label for this instruction.
+    fn label(&self) -> Option<&str>;
+}
+
+/// The parameter list of an instruction.
+#[derive(Clone, Debug)]
+pub enum Parameters<T> {
+    Params(SmallVec<[Param; 3]>),
+    Box {
+        body: T,
+    },
+    ForLoop {
+        indexset: Vec<usize>,
+        loop_param: Option<PyObject>,
+        body: T,
+    },
+    IfElse {
+        true_body: T,
+        false_body: Option<T>,
+    },
+    Switch {
+        cases: Vec<T>,
+    },
+    While {
+        body: T,
+    },
+}
```

Comment:
Is this going to stay like this once #14207 merges?

Reply:
Yeah, this shouldn't need to change. These are different from `ParameterExpression` and the Python-side `Parameter` (which inherits from `ParameterExpression`). 

This is more like an enumeration of the various parameter lists that we support for a given instruction.